Add support for simple keyword replacement.

mostekcm opened the pull request:
There have been some requests for the ability to do a keyword
replacement within a string.  This restricts that ability to do full
JSON replacements, so adding a new way to pass mappings in that allow
for separation of simple and json mappings.

Also added some better error reporting for bad JSON files.  The error
message was previously getting squashed and it wasn't printing the
post-keyword replacement JSON.  Now it prints the error message right,
and adds the post-replacement JSON.  So you can compare pre and post.

@sandrinodimattia which team am I supposed to bug for PR's on this repo now?

Verlic commented:
MD5: c4103f122d27677c9db144cae1394a66

# NSP - Security Report

_This is an automated security audit of this project. Please do not modify its contents_

Vulnerabilities found: **2**
Affected Modules: **uglify-js, uglify-js**
Summary:

- Regular Expression Denial of Service
- Incorrect Handling of Non-Boolean Comparisons During Minification


## Detailed report
#### Regular Expression Denial of Service

Affected Module: **uglify-js**
Installed version: **2.2.5**
Patched versions: **>=2.6.0**
Advisory: https://nodesecurity.io/advisories/48
Path: **auth0-source-control-extension-tools@2.4.3 > auth0-oauth2-express@1.3.0 > jade@1.11.0 > transformers@2.1.0 > uglify-js@2.2.5**

---

#### Incorrect Handling of Non-Boolean Comparisons During Minification

Affected Module: **uglify-js**
Installed version: **2.2.5**
Patched versions: **>= 2.4.24**
Advisory: https://nodesecurity.io/advisories/39
Path: **auth0-source-control-extension-tools@2.4.3 > auth0-oauth2-express@1.3.0 > jade@1.11.0 > transformers@2.1.0 > uglify-js@2.2.5**

---

mostekcm commented:
I also killed a test that no longer makes sense because someone removed the restriction.  Now that get_user is allowed everywhere.

mostekcm commented:
I'm going to redo this a little differently.  Instead of having the different branches in the mappings, I'm just going to change it so that what you put in the file determines whether it is a literal or json replacement.

mostekcm commented:
Squashed and updated to use @@key@@ for json replacements (same behavior as before), but added the ability to do ##key## for a literal replacement.  This makes it so you can do mid-string replacements if needed.

sgmeyer commented on src/utils.js:
```diff
@@ -6,6 +6,11 @@ const ArgumentError = require('auth0-extension-tools').ArgumentError;
 const keywordReplace = function(input, mappings) {
   if (mappings && Object.keys(mappings).length > 0) {
     Object.keys(mappings).forEach(function(key) {
+      const re = new RegExp('##' + key + '##', 'g');
```
Use string interpolation for pattern `##${key}##`.  Otherwise LGTM.

sgmeyer commented:
Had one comment for trivial enhancement, otherwise LGTM.

sgmeyer commented:
Also, after merge and before pushing to npm, let's try to test this with at least one of the other deploy extensions.  GitHub might be good choice for testing.
